lldb/tools/lldb-server/Plugins/AMDGPU/ProcessAMDGPU.cpp

```diff
@@ -86,12 +89,64 @@ lldb::addr_t ProcessAMDGPU::GetSharedLibraryInfoAddress() {
   return LLDB_INVALID_ADDRESS;
 }
 
+ThreadAMDGPU *
+ProcessAMDGPU::FindThread(std::function<bool(ThreadAMDGPU &)> pred) {
+  std::lock_guard<std::recursive_mutex> guard(m_threads_mutex);
+  for (ThreadAMDGPU &thread : AMDGPUThreadRange(m_threads)) {
+    if (pred(thread))
+      return &thread;
+  }
+  return nullptr;
+}
+
+static bool ThreadHasValidStopReason(ThreadAMDGPU &thread) {
+  ThreadStopInfo stop_info;
+  std::string stop_description;
+  thread.GetStopReason(stop_info, stop_description);
+
+  return stop_info.reason != lldb::eStopReasonInvalid &&
+         stop_info.reason != lldb::eStopReasonNone;
+}
```

Comment:
Thinking about this more, we should build this into the `NativeThreadProtocol` class as a method since it uses virtual API calls:
```
bool NativeThreadProtocol::HasValidStopReason() {
  ThreadStopInfo stop_info;
  std::string stop_description;
  if (!GetStopReason(stop_info, stop_description))
    return false;
  return stop_info.reason != lldb::eStopReasonInvalid &&
         stop_info.reason != lldb::eStopReasonNone;
}
```

Reply:
Added to `NativeThreadProtocol`